src/middlewared/middlewared/plugins/apps/rollback.py

```diff
@@ -71,10 +71,13 @@ def rollback(self, job, app_name, options):
             if options['rollback_snapshot'] and (
                 app_volume_ds := self.middleware.call_sync('app.get_app_volume_ds', app_name)
             ):
+                ds = self.middleware.call_sync(
+                    'zfs.resource.query_impl',
+                    {'paths': [app_volume_ds], 'properties': None, 'get_snapshots': True}
+                )
                 snap_name = f'{app_volume_ds}@{options["app_version"]}'
-                if self.middleware.call_sync('zfs.snapshot.query', [['id', '=', snap_name]]):
+                if ds and snap_name in ds[0]['snapshots']:
```

Comment:
I see this code pattern occurs quite a few in your diff. Should we introduce some `zfs.resource.snapshot_exists` helper method to avoid code duplication?